tests/test_zero_data_parallel/test_shard_model_v2.py

```diff
@@ -3,81 +3,69 @@
-        loss = model(data, label)
-    if isinstance(model, ShardedModelV2):
-        model.backward(loss)
-    else:
-        loss.backward()
+from tests.components_to_test.registry import non_distributed_component_funcs
+from common import CONFIG, check_grads_padding, run_fwd_bwd
 
 
-def run_dist(rank, world_size, port):
+def run_dist(rank, world_size, port, use_zero_init_ctx, enable_autocast):
     colossalai.launch(config=CONFIG, rank=rank, world_size=world_size, host='localhost', port=port, backend='nccl')
 
     test_models = ['repeated_computed_layers', 'resnet18', 'bert']
     shard_strategy = TensorShardStrategy()
     for model_name in test_models:
         get_components_func = non_distributed_component_funcs.get_callable(model_name)
-        model, train_dataloader, test_dataloader, optimizer, criterion = get_components_func()
-        model = model(checkpoint=True).half().cuda()
-        zero_model = ShardedModelV2(copy.deepcopy(model), shard_strategy)
-        if dist.get_world_size() > 1:
-            model = DDP(model)
+        model_builder, train_dataloader, _, _, criterion = get_components_func()
+
+        if use_zero_init_ctx:
+            with ZeroInitContext(convert_fp16=True, convert_cuda=True, shard_strategy=shard_strategy, shard_param=True):
+                zero_model = model_builder(checkpoint=True)
+            zero_model = ShardedModelV2(zero_model, shard_strategy)
+
+            model = model_builder(checkpoint=True).half()
+            zero_model.deepcopy_to(model)
+            model = model.cuda()
+        else:
+            model = model_builder(checkpoint=True).half().cuda()
+            zero_model = ShardedModelV2(copy.deepcopy(model), shard_strategy)
+
+        model = DDP(model)
 
         for i, (data, label) in enumerate(train_dataloader):
-            if i > 2:
+            if i > 3:
                 break
 
-            if criterion is None:
-                data, label = data.cuda(), label.cuda()
-                run_fwd_bwd_no_criterion(model, data, label, False)
-                run_fwd_bwd_no_criterion(zero_model, data, label, False)
-            else:
-                data, label = cast_tensor_to_fp16(data).cuda(), label.cuda()
-                run_fwd_bwd(model, data, label, criterion, False)
-                run_fwd_bwd(zero_model, data, label, criterion, False)
+            data, label = cast_tensor_to_fp16(data).cuda(), label.cuda()
+            run_fwd_bwd(model, data, label, criterion, enable_autocast)
+            run_fwd_bwd(zero_model, data, label, criterion, enable_autocast)
 
             check_grads_padding(model, zero_model, loose=True)
 
 
 @pytest.mark.dist
 @pytest.mark.parametrize("world_size", [1, 2, 4])
-def test_shard_model_v2(world_size):
-    run_func = partial(run_dist, world_size=world_size, port=free_port())
+@pytest.mark.parametrize("enable_autocast", [True])
+@pytest.mark.parametrize("use_zero_init_ctx", [True, False])
```

Comment:
This parameter combination can increase the CI time. I feel that the bottleneck is the setup of process groups in most test cases. I shall create our own `parameterize` decorator so that we don't have to spawn multiple processes for each test case. I will accept the current implementation for now.

Reply:
Thanks. I also removed all the existing with_ctx tests. So the time will not change compared to the current develop branch.